x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.test.ts

```diff
@@ -0,0 +1,658 @@
+          expect(params).toEqual(expectedSearchAfterQuery);
+          return sampleDocSearchResultsWithSortId;
+        },
+        alertInstanceFactory: jest.fn(),
+        savedObjectsClient,
+      };
+      const result = await searchAfterAndBulkIndex(
+        sampleDocSearchResultsNoSortIdNoHits,
+        sampleParams,
+        mockService,
+        mockLogger
+      );
+      expect(result).toEqual(true);
+    });
+    test('if successful iteration of while loop with maxDocs and search after returns results with no sort ids', async () => {
+      const searchAfterSortId = '1234567891111';
+      const sampleParams = sampleSignalAlertParams(10);
+      const savedObjectsClient = savedObjectsClientMock.create();
+      const expectedSearchAfterQuery = buildEventsSearchQuery({
+        index: sampleParams.index,
+        from: sampleParams.from,
+        to: sampleParams.to,
+        filter: sampleParams.filter,
+        size: sampleParams.size ? sampleParams.size : 1,
+        searchAfterSortId,
+        maxDocs: sampleParams.maxSignals,
+      });
+      const mockService: AlertServices = {
+        callCluster: async (action: string, params: SearchParams) => {
+          if (action === 'bulk') {
+            expect(action).toEqual('bulk');
+            expect(params.index).toEqual('.siem-signals-10-01-2019');
+            return {
+              took: 100,
+              errors: false,
+              items: [
+                {
+                  fakeItemValue: 'fakeItemKey',
+                },
+              ],
+            };
+          }
+          expect(action).toEqual('search');
+          expect(params.index).toEqual(sampleParams.index);
+          expect(params).toEqual(expectedSearchAfterQuery);
+          return sampleDocSearchResultsNoSortId;
+        },
+        alertInstanceFactory: jest.fn(),
+        savedObjectsClient,
+      };
+      const result = await searchAfterAndBulkIndex(
+        repeatedSearchResultsWithSortId(4),
+        sampleParams,
+        mockService,
+        mockLogger
+      );
+      expect(mockLogger.error).toHaveBeenCalledTimes(1);
+      expect(result).toEqual(true);
+    });
+    test('if logs error when iteration is unsuccessful when bulk index results in a failure', async () => {
+      const searchAfterSortId = '1234567891111';
+      const sampleParams = sampleSignalAlertParams(5);
+      const savedObjectsClient = savedObjectsClientMock.create();
+      const expectedSearchAfterQuery = buildEventsSearchQuery({
+        index: sampleParams.index,
+        from: sampleParams.from,
+        to: sampleParams.to,
+        filter: sampleParams.filter,
+        size: sampleParams.size ? sampleParams.size : 1,
+        searchAfterSortId,
+        maxDocs: sampleParams.maxSignals,
+      });
+
+      const mockService: AlertServices = {
+        callCluster: jest.fn(async (action: string, params: SearchParams) => {
+          if (action === 'bulk') {
```

Comment:
For some of the tests such as this one where the callCluster is called repeatedly with values

I would rely on using if possible just pure jest to ask jest about each time the function was called and with what parameters. Likewise Jest has ways of mocking heterogenous return values:
https://jestjs.io/docs/en/mock-function-api.html#mockfnmockimplementationoncefn

I see what some of these are doing, but the harder part is trying to keep the mocking portion separated from the expect values and avoid mixing them together to where you have part mock and part test pushed together. Rather a good refactor (doesn't have to be with this PR since there is so much here already) is to try and keep them separated from each other and of course keep the expects down to a concentrated 1 or 2 per test.